text/2952-futures.md

```diff
@@ -332,139 +332,68 @@ points」）。固定機制在介紹它的 [RFC](https://github.com/rust-lang/rf
-    system, and this approach also makes things more convoluted when
-    implementing `Future` by hand, for relatively little gain.
-
-The drawback of the "one core trait" approach taken by this RFC is its ergonomic
-hit when writing moveable futures by hand: you now need to import `Pin` and
-`Unpin`, invoke `poll_unpin`, and impl `Unpin` for your types. This is all
-pretty mechanical, but it's a pain. It's possible that improvements in `Pin`
-ergonomics will obviate some of these issues, but there are a lot of open
-questions there still.
-
-On the other hand, a two-trait approach has downsides as well. If we *also*
-remove the error type, there's a combinatorial explosion, since we end up
-needing `Try` variants of each trait (and this extends to related traits, like
-`Stream`, as well). More broadly, with the one-trait approach, `Unpin` acts as a
-kind of "independent knob" that can be applied orthogonally from other concerns;
-with the two-trait approach, it's "mixed in". And both of the two-trait
-approaches run up against compiler limitations at the moment, though of course
-that shouldn't be taken as a deciding factor.
-
-**The primary reason this RFC opts for the one-trait approach is that it's the
-conservative, forward-compatible option, and has proven itself in practice**.
-It's possible to add `MoveFuture`, together with a blanket impl, at any point in the future.
-Thus, starting with just the single `Future` trait as proposed in this RFC keeps our options
-maximally open while we gain experience.
-
-## Rationale, drawbacks and alternatives to the wakeup design (`Waker`)
-
-Previous iterations of this proposal included a separate wakeup type,
-`LocalWaker`, which was `!Send + !Sync` and could be used to implement
-optimized executor behavior without requiring atomic reference counting
-or atomic wakeups. However, in practice, these same optimizations are
-available through the use of thread-local wakeup queues, carrying IDs
-rather than pointers to wakeup objects, and tracking an executor ID or
-thread ID to perform a runtime assertion that a `Waker` wasn't sent across
-threads. For a simple example, a single thread-locked executor with zero
-atomics can be implemented as follows:
+總而言之，futures 的主要替代模型是 callback-base 的方法，在發現當前方法之前嘗試了 callback-base 幾個月。根據我們的經驗，callback 方法在 Rust 中有幾個缺點：
+
+- 它幾乎在所有地方都被強制分配，因此與 no_std 不相容。
+- 它使取消變得**非常**困難，而對於提議的模型，它只是「刪除」。
+- 主觀上，組合器程式碼非常繁瑣，而基於任務的模型則可以輕鬆且快速的達成。
+
+附帶的 [RFC](https://github.com/rust-lang/rfcs/pull/2394) 中提供了整個 async/await 專案的一些其他先備知識和基本原理。
+
+在本節的其餘部分，我們將深入探討特定的 API 設計問題，其中該 RFC 與 futures 0.2 不同。
+
+## 移除內嵌錯誤的基本原理、缺點和替代方法
+
+在主要特徵中刪除內嵌（build-in）錯誤型別有多種原因：
+
+- **改進的類型檢查和推斷**。錯誤型別是當今使用 futures 組合器時最大的痛點之一，無論是試圖讓不同的型別配對，還是在一段代碼無法產生錯誤時導致推理失敗。需要清楚的是，當 `async` 語法可用時，這些問題將變得不那麼明顯。
+
+- **非同步函式**。如果我們保留一個內嵌的錯誤型別，那麼 `async fn` 應該如何工作就不太清楚了：它是否應該總是要求返回的型別是一個 `Result`？如果不是，當返回非 `Result` 型別時會發生什麼？
+
+- **組合器清晰度**。透過它們是否依賴錯誤來拆分組合器可以闡明語義。尤其對於 stream 來說更是如此，其中錯誤處理是常見的混淆來源。
```

Comment:
Not sure if it is better in this, no?

```suggestion
- **組合器清晰度**。按照組合器依賴錯誤與否來拆分之，可以闡明其語義。尤其對於 stream 來說更是如此，其中錯誤處理是常見的混淆來源。
```

Reply:
>Combinator clarity. Splitting up the combinators by whether they rely on errors or not clarifies the semantics. This is especially true for streams, where error handling is a common source of confusion.

I am not sure what is "they" means here.